pkg/authenticator/config/config.go

```diff
@@ -28,12 +29,37 @@ const (
 	DefaultClientCertPath = "/etc/conjur/ssl/client.pem"
 	DefaultTokenFilePath  = "/run/conjur/access-token"
 
+	DefaultConjurVersion = "5"
+
 	// DefaultTokenRefreshTimeout is the default time the system waits to reauthenticate on error
 	DefaultTokenRefreshTimeout = 6 * time.Minute
 )
 
+// NewFromEnvFunc defines the method for creating a new config from the env
+type NewFromEnvFunc func(Options) (*Config, error)
+// ReadSSLCertFunc defines the method for reading an SSL Certificate from the env
+type ReadSSLCertFunc func() ([]byte, error)
+// NewUserNameFunc defines the method for parsing a username from the env
+type NewUserNameFunc func(string) (*Username, error)
+// Options contains the functional options needed for NewFromEnv
+type Options struct {
+	ReadSSLCert ReadSSLCertFunc
+	NewUserName NewUserNameFunc
+}
+
+// DefaultConfigOptions implement the methods defined within config.go
+var DefaultConfigOptions = Options{
+	readSSLCert,
+	NewUsername,
+}
```

Comment:
I could be wrong but I would double-check that you really need any of these types to blackbox-test the authenticator - I think they might not be needed.

Reply:
The issue lies mostly with readSSLCert requiring an actual file to read, and NewUsername is also an external method which requires a specific value, which is difficult to mock out well. If possible I felt this functional option implementation would make it easier to test failure paths in NewFromEnv and to add tests for the other code paths while maintaining readability. Of course, I very much don't want to break existing functionality, but I'll need help figuring out the best way to do this.

Reply:
For now I just swapped the signatures in a quick fix, but I definitely want to hear more about better ways we could do this